// @flashbots/mev-share-client/build/api/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PendingBundle = exports.PendingTransaction = void 0;
class PendingTransaction {
    constructor(event) {
        this.hash = event.hash;
        this.logs = event.logs || undefined;
        this.to = event.txs && event.txs[0].to;
        this.functionSelector = event.txs && event.txs[0].functionSelector;
        this.callData = event.txs && event.txs[0].callData;
        this.gasUsed = event.gasUsed ? BigInt(event.gasUsed) : undefined;
        this.mevGasPrice = event.mevGasPrice ? BigInt(event.mevGasPrice) : undefined;
    }
}
exports.PendingTransaction = PendingTransaction;
class PendingBundle {
    constructor(event) {
        this.hash = event.hash;
        this.logs = event.logs || undefined;
        this.txs = event.txs;
        this.gasUsed = event.gasUsed ? BigInt(event.gasUsed) : undefined;
        this.mevGasPrice = event.mevGasPrice ? BigInt(event.mevGasPrice) : undefined;
    }
}
exports.PendingBundle = PendingBundle;
